In [1]:
import pandas as pd

In [2]:
df = pd.read_csv(r'C:\Users\Laxman\OneDrive\Desktop\Ad click data.csv')

In [3]:
df.head()

VisitorID  TimeSpent  Age  AvgIncome  InternetUsage     AdTopic  \
0    5183153      87.97   43   55901.12         185.46  product_11   
1    4023265      51.63   50   39132.00         176.73   product_8   
2    4708083      82.37   38   57032.36         210.60   product_6   
3    9771815      62.06   45   48868.00         190.05  product_19   
4    6451317      77.66   31   61608.23         204.86  product_11   

                CountryName CityCode Male TimePeriod   Weekday    Month  Year  \
0                    Serbia   City_5   No  Mid-Night  Thursday     July  2020   
1              Turkmenistan   City_1   No    Evening  Saturday     June  2020   
2  Northern Mariana Islands   City_2   No    Morning   Tuesday  January  2020   
3              South Africa   City_3  Yes    Morning  Thursday    April  2020   
4                Guadeloupe   City_2   No       Noon  Thursday  January  2020   

   Clicked  
0        0  
1        1  
2        0  
3        1  
4        0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6657 entries, 0 to 6656
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   VisitorID      6657 non-null   int64  
 1   TimeSpent      6657 non-null   float64
 2   Age            6657 non-null   int64  
 3   AvgIncome      6657 non-null   float64
 4   InternetUsage  6657 non-null   float64
 5   AdTopic        6657 non-null   object 
 6   CountryName    6657 non-null   object 
 7   CityCode       6657 non-null   object 
 8   Male           6657 non-null   object 
 9   TimePeriod     6657 non-null   object 
 10  Weekday        6657 non-null   object 
 11  Month          6657 non-null   object 
 12  Year           6657 non-null   int64  
 13  Clicked        6657 non-null   int64  
dtypes: float64(3), int64(4), object(7)
memory usage: 728.2+ KB


In [5]:
df.isnull().sum()

VisitorID        0
TimeSpent        0
Age              0
AvgIncome        0
InternetUsage    0
AdTopic          0
CountryName      0
CityCode         0
Male             0
TimePeriod       0
Weekday          0
Month            0
Year             0
Clicked          0
dtype: int64

In [6]:
df_dummy = pd.get_dummies(df,columns=['AdTopic','CountryName','CityCode','Male','TimePeriod','Weekday','Month'],drop_first=True)

In [7]:
df_dummy.head()

VisitorID  TimeSpent  Age  AvgIncome  InternetUsage  Year  Clicked  \
0    5183153      87.97   43   55901.12         185.46  2020        0   
1    4023265      51.63   50   39132.00         176.73  2020        1   
2    4708083      82.37   38   57032.36         210.60  2020        0   
3    9771815      62.06   45   48868.00         190.05  2020        1   
4    6451317      77.66   31   61608.23         204.86  2020        0   

   AdTopic_product_10  AdTopic_product_11  AdTopic_product_12  ...  \
0                   0                   1                   0  ...   
1                   0                   0                   0  ...   
2                   0                   0                   0  ...   
3                   0                   0                   0  ...   
4                   0                   1                   0  ...   

   Weekday_Sunday  Weekday_Thursday  Weekday_Tuesday  Weekday_Wednesday  \
0               0                 1                0                  0   
1               0                 0                0                  0   
2               0                 0                1                  0   
3               0                 1                0                  0   
4               0                 1                0                  0   

   Month_February  Month_January  Month_July  Month_June  Month_March  \
0               0              0           1           0            0   
1               0              0           0           1            0   
2               0              1           0           0            0   
3               0              0           0           0            0   
4               0              1           0           0            0   

   Month_May  
0          0  
1          0  
2          0  
3          0  
4          0  

[5 rows x 298 columns]

In [9]:
df_dummy.shape

(6657, 298)

Identify the input and output features

In [10]:
y = df_dummy[['Clicked']]
x = df_dummy.drop(columns=['Clicked'])

Lets perform feature scaling on the data using Standardisation

In [11]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_scal = sc.fit_transform(x)

In [12]:
x_scal = pd.DataFrame(x_scal,columns=x.columns)
x_scal.head()

VisitorID  TimeSpent       Age  AvgIncome  InternetUsage  Year  \
0  -0.138270   1.361862  0.522180  -0.002240       0.011863   0.0   
1  -0.585053  -0.981368  1.158854  -1.281412      -0.190283   0.0   
2  -0.321265   1.000770  0.067412   0.084052       0.593987   0.0   
3   1.629259  -0.308834  0.704087  -0.538736       0.118146   0.0   
4   0.350220   0.697066 -0.569262   0.433106       0.461076   0.0   

   AdTopic_product_10  AdTopic_product_11  AdTopic_product_12  \
0           -0.190024            5.460718           -0.187892   
1           -0.190024           -0.183126           -0.187892   
2           -0.190024           -0.183126           -0.187892   
3           -0.190024           -0.183126           -0.187892   
4           -0.190024            5.460718           -0.187892   

   AdTopic_product_13  ...  Weekday_Sunday  Weekday_Thursday  Weekday_Tuesday  \
0             -0.1963  ...       -0.415235          2.487767        -0.404233   
1             -0.1963  ...       -0.415235         -0.401967        -0.404233   
2             -0.1963  ...       -0.415235         -0.401967         2.473820   
3             -0.1963  ...       -0.415235          2.487767        -0.404233   
4             -0.1963  ...       -0.415235          2.487767        -0.404233   

   Weekday_Wednesday  Month_February  Month_January  Month_July  Month_June  \
0            -0.4002       -0.406494      -0.417470    2.481545   -0.405741   
1            -0.4002       -0.406494      -0.417470   -0.402975    2.464628   
2            -0.4002       -0.406494       2.395382   -0.402975   -0.405741   
3            -0.4002       -0.406494      -0.417470   -0.402975   -0.405741   
4            -0.4002       -0.406494       2.395382   -0.402975   -0.405741   

   Month_March  Month_May  
0    -0.402975  -0.407998  
1    -0.402975  -0.407998  
2    -0.402975  -0.407998  
3    -0.402975  -0.407998  
4    -0.402975  -0.407998  

[5 rows x 297 columns]

Split the data into train and test

In [13]:
from sklearn.model_selection import train_test_split

In [14]:
x_train,x_test,y_train,y_test = train_test_split(x_scal,y,test_size=0.2,random_state=42)

In [15]:
len(x_train),len(x_test),len(y_train),len(y_test)

(5325, 1332, 5325, 1332)

## Building Logistic Regression

In [16]:
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression()
model = log_reg.fit(x_train,y_train)

C:\Users\Laxman\anaconda3\lib\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [17]:
y_test.head()

Clicked
3332        1
217         0
4715        0
2939        1
1544        1

In [19]:
y_test['Prediction_LR'] = model.predict(x_test)

In [20]:
y_test.head()

Clicked  Prediction_LR
3332        1              1
217         0              0
4715        0              0
2939        1              1
1544        1              1

In [21]:
from sklearn.metrics import accuracy_score,confusion_matrix

In [22]:
print(confusion_matrix(y_test['Clicked'],y_test['Prediction_LR']))

[[679  37]
 [ 63 553]]


In [23]:
print(accuracy_score(y_test['Clicked'],y_test['Prediction_LR']))

0.924924924924925


The Logistic Regression model is able to predict the data with 92.4% accuracy. This is a good value

In [24]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()
model = dt.fit(x_train,y_train)

In [25]:
y_test['Prediction_DT'] = model.predict(x_test)

In [28]:
y_test.head()

Clicked  Prediction_LR  Prediction_DT
3332        1              1              1
217         0              0              0
4715        0              0              0
2939        1              1              1
1544        1              1              0

In [27]:
print(confusion_matrix(y_test['Clicked'],y_test['Prediction_DT']))

[[679  37]
 [ 75 541]]


In [29]:
print(accuracy_score(y_test['Clicked'],y_test['Prediction_DT']))

0.9159159159159159


The Decision tree algorithm is not performing better than logistic regression and it is giving an accuracy of 91.6%